Fixes #23938 - api/tasks uses the index-layout

sharvit opened the pull request:
Refactor the `api/tasks_controller#index` so it will based on the basic controllers

`sort_by` and `sort_order` won't work anymore and you will need to use `order=id desc` instead.
is that a problem? should we show a deprecation message? should migrate other places?
I just not sure where it got used, I know about `index page` and the `breadcrumbs`. (hammer?)

sharvit commented:
Related to https://github.com/theforeman/foreman-tasks/pull/330#discussion_r194354145

adamruzicka commented:
Just a couple of observations

Timestamps are formatted differently, I don't feel strongly about this.
```
old: "started_at": "2018-06-08T08:20:03.000Z"
new: "started_at": "2018-06-08 08:20:03 UTC"
```

We used to order `started_at DESC` by default, it would be nice if you could keep doing that.

Would it be possible to still support using `sort_by` and `sort_order` and print a deprecation when someone would use it? It would give us a bit of breathing space to catch all the places where it was being used without causing breakage.

adamruzicka commented:
@sharvit could you take a look at this again when you have a bit of spare time?

sharvit commented:
> Timestamps are formatted differently, I don't feel strongly about this.
> 
> old: "started_at": "2018-06-08T08:20:03.000Z"
> new: "started_at": "2018-06-08 08:20:03 UTC"

I just don't get it, I switched the `index.rabl` to reuse the `show.rabl` which used before and it still the same, I can't find any logic differences when it comes to dates formating :/ 

> We used to order started_at DESC by default, it would be nice if you could keep doing that.

Added `default_scope` to the task model, do you think it's a good practice?

> Would it be possible to still support using sort_by and sort_order and print a deprecation when someone would use it? It would give us a bit of breathing space to catch all the places where it was being used without causing breakage.

Added support for the old sorting params with a deprecation warning. See `def search_options`

sharvit commented:
@adamruzicka Can you rereview please?

sharvit commented on app/models/foreman_tasks/task.rb:
```diff
@@ -33,6 +33,8 @@ def check_permissions_after_save
     has_many :owners, -> { where(['foreman_tasks_locks.name = ?', Lock::OWNER_LOCK_NAME]) },
              :through => :locks, :source => :resource, :source_type => 'User'
 
+    default_scope -> { order(start_at: :desc) }
```
It seems to be an issue in foreman, it is consistent with other API routes in foreman.

See:
`/api/v2/hosts`
`/api/v2/architectures`

Issue created: https://projects.theforeman.org/issues/25616

sharvit commented on app/controllers/foreman_tasks/api/tasks_controller.rb:
```diff
@@ -109,44 +109,25 @@ def bulk_resume
+        if params[:sort_by] || params[:sort_order]
+          Foreman::Deprecation.api_deprecation_warning(
+            "The sort params sort_by and sort_order are deprecated.
+            Please use the order param instead as onc string 'params=started_at desc'"
```
Fixed

sharvit commented:
np @adamruzicka :+1: 
can you revisit?

sharvit commented:
Thanks @adamruzicka, fixed.

adamruzicka commented:
Could you please rebase so the tests can pass?

sharvit commented:
Sure @adamruzicka, rebased.

adamruzicka commented:
I looked into this and the test failures are actually related. How about creating a named scope instead of the default one (for a lot of task queries we don't need the ordering at all) and using that named scope in the api tasks controller?

sharvit commented:
@adamruzicka I'm sorry but I don't have the capacity to continue working on it.
Can we close it?
Or maybe someone else can continue?

bkearney commented:
@adamruzicka @sharvit Seems pretty old, can we get this merged?

sharvit commented:
As I discussed with @adamruzicka offline, he will continue the effort.
closing.

I am keeping the branch open: https://github.com/sharvit/foreman-tasks/tree/refactor/api-tasks-index-rabl
